gengokit/template/NAME-service/NAME-cli-client/client_submain.gotemplate

```diff
@@ -0,0 +1,126 @@
+package main
+
+import (
+	"encoding/json"
+	"flag"
+	"fmt"
+	"os"
+	"strconv"
+	"strings"
+	"time"
+
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+
+	"github.com/pkg/errors"
+
+	// This Service
+	clientHandler "{{.ImportPath -}} /generated/cli/handlers"
+	grpcclient "{{.ImportPath -}} /generated/client/grpc"
+	httpclient "{{.ImportPath -}} /generated/client/http"
+	pb "{{.PBImportPath -}}"
+)
+
+var (
+	_ = strconv.ParseInt
+	_ = strings.Split
+	_ = json.Compact
+	_ = errors.Wrapf
+	_ = pb.Register{{.Service.Name}}Server
+)
+
+// submain exists to act as the functional main, but will return exit codes to
+// the actual main instead of calling os.Exit directly. This is done to allow
+// the defered functions to be called, since if os.Exit where called directly
+// from this function, none of the defered functions set up by this function
+// would be called.
+func submain() int {
```

Comment:
Calling it `run` seems like using `self` as a method receiver; overly generic, conveying nothing of the purposes this code is meant to fulfill. Nearly any function could accurately be called `run`.

However, it also would violate the convention of a `run` package described in the blog post you linked above which describes a `run` package as (emphasis mine):

> The run package contains the meat of the logic of your binary. You should write this package as if it were a standalone library. *It should be far removed from any thoughts of CLI, flags, etc. It should take in structured data and return errors.*

Since the structure and purpose of the convention we're using in this PR is so narrow and so focused, I think that a name of `main` and `submain` with the explanatory comments is the way to go.